uniffi_macros/src/record.rs

```diff
@@ -0,0 +1,96 @@
+use proc_macro2::{Span, TokenStream};
+use quote::quote;
+use syn::{Data, DeriveInput};
+use uniffi_meta::{FieldMetadata, RecordMetadata};
+
+use crate::{export::metadata::convert::convert_type, util::create_metadata_static_var};
+
+pub fn expand_record(input: DeriveInput, module_path: Vec<String>) -> TokenStream {
+    let fields = match input.data {
+        Data::Struct(s) => Some(s.fields),
+        _ => None,
+    };
+
+    let ident = &input.ident;
+
+    let (write_impl, try_read_fields) = match &fields {
+        Some(fields) => fields
+            .iter()
+            .map(|f| {
+                let ident = &f.ident;
+                let ty = &f.ty;
+
+                let write_field = quote! {
+                    <#ty as ::uniffi::FfiConverter>::write(obj.#ident, buf);
+                };
+                let try_read_field = quote! {
+                    #ident: <#ty as ::uniffi::FfiConverter>::try_read(buf)?,
```

Comment:
I like this as a first step.  I guess the next step would be getting it to work with user types, where the type doesn't implement `FfiConverter` itself.  I think that goes back to an issue that we discussed very early on.

Do you have a plan for this?  If not, I could spend some time trying to get it work if that would be helpful.

Reply:
Yes, this goes back to what we discussed pretty early on. I would still prefer to not support "external" ffi-converter types, and instead encourage the newtype pattern where people were previously implementing `FfiConverter` with a `RustType` other than `Self`.

For the multi-crate setups I'm thinking about, neither would be required though; there's also the option of types having `#[cfg_attr(feature = "uniffi", derive(uniffi::Record))]` so a crate can *optionally* have a `FfiConverter` impl in an upstream crate.

Reply:
I do think that almost all of the time it's going to be one of those multi-crate setups where you control all the crates.

For completely external types, I can be convinced that requiring the newtype pattern is the best way to go.  I'm mostly just concerned that there's some use case out there where this would be lead to very bad ergonomics.  I'm going to put the question out there on `#uniffi` and see if anyone has a comment.